### PR TITLE
fix(mpc): precise resampling around ego for stabling control

### PR DESCRIPTION
**Pull request by takayuki5168**
## Description

Without the PR, the relative time vector around ego changes over time, resulting in pretty bad control.
https://tier4.atlassian.net/browse/RT0-28891?focusedCommentId=137957 (Videos in internal link)

This PR fixes the issue by
- 1. make dynamicSmoothingVelocity use the segment index of ego pose which may be the previous of nearest index.
- 2. make resampleMPCTrajectoryByDistance resample the ego pose projected to the trajectory
    - This makes a major improvement on the issue.
- 3. make calcNearestPoseInterp stable over time by using sequential points not by using skipping two points

image

The implementation (with this PR and even without this PR) looks hard to read and maintain. Need to be refactored.

## Tests performed

psim

## Effects on system behavior

Stabler steering control

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [x] The PR follows the [pull request guidelines].

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

File: control/mpc_lateral_controller/src/mpc_utils.cpp [hunk 1]
```diff
@@ -92,7 +106,16 @@ std::pair<bool, MPCTrajectory> resampleMPCTrajectoryByDistance(
   }
 
   std::vector<double> output_arclength;
-  for (double s = 0; s < input_arclength.back(); s += resample_interval_dist) {
+  for (double s = std::clamp(
```
Comment (TakaHoribe) [inline control/mpc_lateral_controller/src/mpc_utils.cpp:109]:
```suggestion
  // To accurately sample the ego point, resample separately in the forward direction and the
  // backward direction from the current position.
  for (double s = std::clamp(
```

File: control/mpc_lateral_controller/src/mpc_utils.cpp [hunk 2]
```diff
@@ -92,7 +106,16 @@ std::pair<bool, MPCTrajectory> resampleMPCTrajectoryByDistance(
...
+         input_arclength.at(nearest_seg_idx) + ego_offset_to_segment, 0.0,
+         input_arclength.back() - 1e-6);
+       0 <= s; s -= resample_interval_dist) {
+    output_arclength.push_back(s);
+  }
+  std::reverse(output_arclength.begin(), output_arclength.end());
+  for (double s = std::max(input_arclength.at(nearest_seg_idx) + ego_offset_to_segment, 0.0) +
```
Comment (TakaHoribe) [inline control/mpc_lateral_controller/src/mpc_utils.cpp:116]: is `input_arclength.at(nearest_seg_idx) + ego_offset_to_segment` resampled twice?

File: control/mpc_lateral_controller/src/mpc_utils.cpp [hunk 3]
```diff
@@ -274,13 +297,16 @@ bool calcMPCTrajectoryTime(MPCTrajectory & traj)
 }
 
 void dynamicSmoothingVelocity(
-  const size_t start_idx, const double start_vel, const double acc_lim, const double tau,
+  const size_t start_seg_idx, const double start_vel, const double acc_lim, const double tau,
   MPCTrajectory & traj)
 {
   double curr_v = start_vel;
-  traj.vx.at(start_idx) = start_vel;
+  traj.vx.at(start_seg_idx) = start_vel;
```
Comment (TakaHoribe) [inline control/mpc_lateral_controller/src/mpc_utils.cpp:304]:
```suggestion
  // set current velocity in both start and end point of the segment
  traj.vx.at(start_seg_idx) = start_vel;
```

File: control/mpc_lateral_controller/src/mpc_utils.cpp (hunk 2 above)
Comment (takayuki5168) [reply control/mpc_lateral_controller/src/mpc_utils.cpp:118]: No, this case uses `double s = std::max(input_arclength.at(nearest_seg_idx) + ego_offset_to_segment, 0.0) + "resample_interval_dist"`.

**Review by TakaHoribe — APPROVED** [review]
Confirmed it works well in psim.

**Pull request merged**